sklearn/impute/_knn.py

```diff
@@ -115,6 +117,15 @@ class KNNImputer(_BaseImputer):
            [8. , 8. , 7. ]])
     """
 
+    _parameter_constraints = {
+        "missing_values": [Real, Integral, str, None],
+        "n_neighbors": [Interval(Integral, 1, None, closed="left")],
+        "weights": [StrOptions({"uniform", "distance"}), callable],
+        "metric": [StrOptions(set(_NAN_METRICS)), callable],
+        "copy": ["boolean"],
+        "add_indicator": ["boolean"],
```

Comment:
The `missing_values` and `add_indicator` should be validated in the base class `_BaseImputer` in `sklearn/imputer/base.py`.

Here, we will just get the dictionary and add up the missing constraints that are specific to `KNNImputer`.

Reply:
As an example of validating in a base class and then adding the specific constraints in the specified class, you can have a look at this PR: https://github.com/scikit-learn/scikit-learn/pull/23773/files#diff-72625099b4451eed49760f66ac5a2665b7c0b799b21c55d3beed0bac43a3ef28R450-R454

The pattern is similar with a base class for AdaBoost and then specific class for the classifier and regressor.

Reply:
I moved some of the parameters to the base class, but I did not add a call to the `_validate_params` in the `_BaseImputer.fit()`,
because there are more parameters that need to be validated and probably added to the dictionary.